pacman/operations/routing_info_allocator_algorithms/global_zoned_routing_info_allocator.py

```diff
@@ -0,0 +1,176 @@
+        self._application_graph = application_graph
+        self._machine_graph = machine_graph
+        self._placements = placements
+        self._n_keys_map = n_keys_map
+
+        check_algorithm_can_support_constraints(
+            constrained_vertices=machine_graph.outgoing_edge_partitions,
+            supported_constraints=[ContiguousKeyRangeContraint],
+            abstract_constraint_type=AbstractKeyAllocatorConstraint)
+
+        self._calculate_zones()
+        return self._simple_allocate()
+
+    def _calculate_zones(self):
+        progress = ProgressBar(
+            self._application_graph.n_vertices, "Calculating zones")
+        max_partitions = 0
+        max_machine_vertices = 0
+        max_keys = 0
+        n_app_vertices = 0
+
+        for app_vertex in progress.over(self._application_graph.vertices):
+            app_max_partitions = 0
+            n_machine_vertices = 0
+            for vertex in app_vertex.machine_vertices:
+                partitions = self._machine_graph.\
+                    get_outgoing_edge_partitions_starting_at_vertex(vertex)
+                n_partitions = 0
+                local_max_keys = 0
+                for partition in partitions:
+                    if partition.traffic_type == EdgeTrafficType.MULTICAST:
+                        n_keys = self._n_keys_map.n_keys_for_partition(
+                            partition)
+                        local_max_keys = max(local_max_keys, n_keys)
+                        n_partitions += 1
+                app_max_partitions = max(app_max_partitions, n_partitions)
+                n_machine_vertices += 1
+            if local_max_keys > 0:
+                max_partitions = max(max_partitions, app_max_partitions)
+                max_machine_vertices = max(
+                    n_machine_vertices, max_machine_vertices)
+                max_keys = max(local_max_keys, max_keys)
+                n_app_vertices += 1
+
+        self._n_bits_machine = _bits_needed(max_machine_vertices)
+        self._n_bits_partition = _bits_needed(max_partitions)
+        self._n_bits_atoms = _bits_needed(max_keys)
+        self._n_bits_total = sum(
+            self._n_bits_machine, self._n_bits_partition, self._n_bits_atoms)
+        n_bits_vertices = _bits_needed(n_app_vertices)
+
+        if (self._n_bits_total + n_bits_vertices) > BITS_IN_KEY:
+            raise PacmanRouteInfoAllocationException(
+                "Unable to use GlobalZonedRoutingInfoAllocator as it needs "
+                "{} + {} + {} + {} bits".format(
+                    self._n_bits_machine, self._n_bits_partition,
+                    self._n_bits_atoms, n_bits_vertices))
+        return max_partitions
+
+    def _simple_allocate(self):
+        progress = ProgressBar(
+            self._application_graph.n_vertices, "Allocating routing keys")
+        routing_infos = RoutingInfo()
+        by_app_vertex = dict()
+        mask = 0xFFFFFFFF - ((2 ** self._n_bits_atoms) - 1)
```

Comment:
magic number. used twice. named constant?

Reply:
also why isnt that in pacman constants. you'd assume we use that elsewhere............

Reply:
oh its in pacman constants! USE THE NAMED CONSTANT!